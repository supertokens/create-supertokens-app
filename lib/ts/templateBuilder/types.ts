// types/config.ts
export type ConfigType =
    | "all_auth"
    | "emailpassword"
    | "multifactorauth"
    | "multitenancy"
    | "passwordless"
    | "thirdparty" // Added for thirdparty-only config
    | "thirdpartyemailpassword"
    | "thirdpartypasswordless";

export type IndividualRecipe =
    | "emailPassword"
    | "thirdParty"
    | "passwordless"
    | "session"
    | "dashboard"
    | "userRoles"
    | "multiFactorAuth"
    | "accountLinking"
    | "emailVerification"
    | "totp"
    | "multitenancy";

export type FrontendFramework = "react" | "vue" | "angular" | "solid";

export type Language = "ts" | "py" | "go";

export type AppInfo = {
    appName: string;
    apiDomain: string;
    websiteDomain: string;
    apiBasePath: string;
    websiteBasePath: string;
    connectionURI: string;
};

export type OAuthProvider = {
    id: string;
    clientId: string;
    clientSecret: string;
    additionalConfig?: Record<string, any>;
};
