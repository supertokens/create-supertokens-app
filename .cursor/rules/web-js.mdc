---
description: 
globs: 
alwaysApply: false
---
## WebJS

| factor          | webjs imports                                            | webjs inits                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |
| --------------- | -------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `emailpassword` | import SuperTokens from "supertokens-web-js";            | (window as any).supertokensUIEmailPassword.init(),                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
| `thirdparty`    | import SuperTokens from "supertokens-web-js";            | (window as any).supertokensUIThirdParty.init({<br>                signInAndUpFeature: {<br>                    providers: [<br>                        (window as any).supertokensUIThirdParty.Github.init(),<br>                        (window as any).supertokensUIThirdParty.Google.init(),<br>                        (window as any).supertokensUIThirdParty.Apple.init(),<br>                        (window as any).supertokensUIThirdParty.Twitter.init(),<br>                    ],<br>                },<br>            }), |
| `link_email`    | import Session from "supertokens-web-js/recipe/session"; | (window as any).supertokensUIPasswordless.init({<br>                contactMethod: "EMAIL_OR_PHONE",<br>            }),                                                                                                                                                                                                                                                                                                                                                                                                                |
| `link_phone`    | import Session from "supertokens-web-js/recipe/session"; | (window as any).supertokensUIPasswordless.init({<br>                contactMethod: "EMAIL_OR_PHONE",<br>            }),                                                                                                                                                                                                                                                                                                                                                                                                                |
| `otp_phone`     | import Session from "supertokens-web-js/recipe/session"; | (window as any).supertokensUIPasswordless.init({<br>                contactMethod: "EMAIL_OR_PHONE",<br>            }),                                                                                                                                                                                                                                                                                                                                                                                                                |
| `otp_email`     | import Session from "supertokens-web-js/recipe/session"; | (window as any).supertokensUIPasswordless.init({<br>                contactMethod: "EMAIL_OR_PHONE",<br>            }),                                                                                                                                                                                                                                                                                                                                                                                                                |
| `totp`          | import SuperTokens from "supertokens-web-js";            | (window as any).supertokensUITOTP.init(),<br>            (window as any).supertokensUISession.init(),                                                                                                                                                                                                                                                                                                                                                                                                                                  |

**Notes**: 
- Session, Dashboard, UserRoles also imported for all factors, in addition to their own config.
- if both link_email and link_phone are present as factors, the contactMethod in init becomes "EMAIL_OR_PASSWORD".
- if both otp_email and otp_phone are present as factors, the flowType in init becomes "USER_INPUT_CODE_AND_MAGIC_LINK".
- If a second factor is present, always include the MultiFactorAuth recipe.
- Second factors specific behavior always go in the override function of the MultiFactorAuth recipe!

### Examples

#### All auth

```ts
import SuperTokens from "supertokens-web-js";
import Session from "supertokens-web-js/recipe/session";

export function initSuperTokensUI() {
    (window as any).supertokensUIInit("supertokensui", {
        appInfo: {
            websiteDomain: "http://localhost:3000",
            apiDomain: "http://localhost:3001",
            appName: "SuperTokens Demo App",
        },
        recipeList: [
            (window as any).supertokensUIEmailPassword.init(),
            (window as any).supertokensUIThirdParty.init({
                signInAndUpFeature: {
                    providers: [
                        (window as any).supertokensUIThirdParty.Github.init(),
                        (window as any).supertokensUIThirdParty.Google.init(),
                        (window as any).supertokensUIThirdParty.Apple.init(),
                        (window as any).supertokensUIThirdParty.Twitter.init(),
                    ],
                },
            }),
            (window as any).supertokensUIPasswordless.init({
                contactMethod: "EMAIL_OR_PHONE",
            }),
            (window as any).supertokensUISession.init(),
        ],
    });
}

export function initSuperTokensWebJS() {
    SuperTokens.init({
        appInfo: {
            appName: "SuperTokens Demo App",
            apiDomain: "http://localhost:3001",
        },
        recipeList: [Session.init()],
    });
}
```

#### Multi-factor auth

```ts
import SuperTokens from "supertokens-web-js";
import Session from "supertokens-web-js/recipe/session";
import EmailVerification from "supertokens-web-js/recipe/emailverification";
import MultiFactorAuth from "supertokens-web-js/recipe/multifactorauth";

export function initSuperTokensUI() {
    (window as any).supertokensUIInit("supertokensui", {
        appInfo: {
            websiteDomain: "http://localhost:3000",
            apiDomain: "http://localhost:3001",
            appName: "SuperTokens Demo App",
        },
        recipeList: [
            (window as any).supertokensUIEmailPassword.init(),
            (window as any).supertokensUIThirdParty.init({
                signInAndUpFeature: {
                    providers: [
                        (window as any).supertokensUIThirdParty.Github.init(),
                        (window as any).supertokensUIThirdParty.Google.init(),
                        (window as any).supertokensUIThirdParty.Apple.init(),
                        (window as any).supertokensUIThirdParty.Twitter.init(),
                    ],
                },
            }),
            (window as any).supertokensUIPasswordless.init({
                contactMethod: "EMAIL_OR_PHONE",
            }),
            (window as any).supertokensUIEmailVerification.init({
                mode: "REQUIRED",
            }),
            (window as any).supertokensUIMultiFactorAuth.init({
                firstFactors: ["thirdparty", "emailpassword"],
            }),
            (window as any).supertokensUITOTP.init(),
            (window as any).supertokensUISession.init(),
        ],
    });
}

export function initSuperTokensWebJS() {
    SuperTokens.init({
        appInfo: {
            appName: "SuperTokens Demo App",
            apiDomain: "http://localhost:3001",
        },
        recipeList: [Session.init(), EmailVerification.init(), MultiFactorAuth.init()],
    });
}
```

These rules and examples apply only to [template.ts](mdc:boilerplate/frontend/shared/web-js/template.ts) - and nowhere outside of it.


