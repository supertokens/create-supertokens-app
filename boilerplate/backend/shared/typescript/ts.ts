import { type OAuthProvider, type ConfigType } from "../../../../lib/ts/templateBuilder/types";
import { configToRecipes } from "../../../../lib/ts/templateBuilder/constants";
import { config } from "../../../shared/config/base";
import { getAppInfo } from "../../../shared/config/appInfo";
import { oAuthProviders } from "../../../backend/shared/config/oAuthProviders";
import { UserFlags } from "../../../../lib/ts/types";

interface TypeScriptTemplateOptions {
    configType: ConfigType;
    userArguments?: UserFlags;
    isFullStack?: boolean; // Added isFullStack flag
}

export const tsRecipeImports = {
    emailPassword: 'import EmailPassword from "supertokens-node/recipe/emailpassword";',
    thirdParty:
        'import ThirdParty from "supertokens-node/recipe/thirdparty";\nimport type { ProviderInput } from "supertokens-node/recipe/thirdparty/types";', // Use import type
    passwordless: 'import Passwordless from "supertokens-node/recipe/passwordless";',
    session: 'import Session from "supertokens-node/recipe/session";',
    dashboard: 'import Dashboard from "supertokens-node/recipe/dashboard";',
    userRoles: 'import UserRoles from "supertokens-node/recipe/userroles";',
    multiFactorAuth: 'import MultiFactorAuth from "supertokens-node/recipe/multifactorauth";',
    accountLinking: 'import AccountLinking from "supertokens-node/recipe/accountlinking";',
    emailVerification: 'import EmailVerification from "supertokens-node/recipe/emailverification";',
    totp: 'import TOTP from "supertokens-node/recipe/totp";',
    multitenancy: 'import Multitenancy from "supertokens-node/recipe/multitenancy";',
} as const;

export const tsRecipeInits = {
    emailPassword: () => `EmailPassword.init()`,
    thirdParty: (providers: OAuthProvider[]) => `ThirdParty.init({
            signInAndUpFeature: {
                providers: [
                    ${providers
                        .map(
                            (provider) => `{
                        config: {
                            thirdPartyId: "${provider.id}",
                            clients: [
                                {
                                    clientId: "${provider.clientId}",
                                    clientSecret: "${provider.clientSecret}",
                                    ${
                                        provider.additionalConfig
                                            ? `additionalConfig: ${JSON.stringify(provider.additionalConfig, null, 36)}`
                                            : ""
                                    }
                                },
                            ],
                        },
                    }`
                        )
                        .join(",\n                    ")}
                ],
            },
        })`,
    passwordless: (userArguments?: UserFlags) => {
        // Determine contact method and flow type based on user arguments
        let contactMethod = "EMAIL";
        let flowType; // To handle OTP and magic link combined flows

        const hasLinkEmail =
            userArguments?.firstfactors?.includes("link-email") || userArguments?.secondfactors?.includes("link-email");
        const hasLinkPhone =
            userArguments?.firstfactors?.includes("link-phone") || userArguments?.secondfactors?.includes("link-phone");
        const hasOtpEmail =
            userArguments?.firstfactors?.includes("otp-email") || userArguments?.secondfactors?.includes("otp-email");
        const hasOtpPhone =
            userArguments?.firstfactors?.includes("otp-phone") || userArguments?.secondfactors?.includes("otp-phone");

        // Determine contact method
        if ((hasLinkEmail || hasOtpEmail) && (hasLinkPhone || hasOtpPhone)) {
            contactMethod = "EMAIL_OR_PHONE";
        } else if (hasLinkPhone || hasOtpPhone) {
            contactMethod = "PHONE";
        } else {
            contactMethod = "EMAIL";
        }

        // Determine flow type
        const hasLinkFactor = hasLinkEmail || hasLinkPhone;
        const hasOtpFactor = hasOtpEmail || hasOtpPhone;

        if (hasLinkFactor && hasOtpFactor) {
            flowType = "USER_INPUT_CODE_AND_MAGIC_LINK";
        } else if (hasLinkFactor) {
            flowType = "MAGIC_LINK";
        } else if (hasOtpFactor) {
            flowType = "USER_INPUT_CODE";
        }

        return `Passwordless.init({
    contactMethod: "${contactMethod}",${flowType ? `\n    flowType: "${flowType}"` : ""}
})`;
    },
    session: () => `Session.init()`,
    dashboard: () => `Dashboard.init()`,
    userRoles: () => `UserRoles.init()`,
    multiFactorAuth: (firstFactors?: string[], secondFactors?: string[]) => {
        const firstFactorsStr = (firstFactors || ["thirdparty", "emailpassword"])
            .map((factor) => `"${factor}"`)
            .join(", ");

        let init = `MultiFactorAuth.init({
    firstFactors: [${firstFactorsStr}]`;

        if (secondFactors && secondFactors.length > 0) {
            const factorMapping: Record<string, string> = {
                totp: "TOTP",
                "otp-email": "OTP_EMAIL",
                "otp-phone": "OTP_PHONE",
                "link-email": "LINK_EMAIL",
                "link-phone": "LINK_PHONE",
                otp_email: "OTP_EMAIL",
                otp_phone: "OTP_PHONE",
                link_email: "LINK_EMAIL",
                link_phone: "LINK_PHONE",
            };

            const factorIds: string[] = [];
            secondFactors.forEach((factor) => {
                if (factorMapping[factor]) {
                    factorIds.push(`MultiFactorAuth.FactorIds.${factorMapping[factor]}`);
                }
            });

            if (factorIds.length > 0) {
                init += `,
    override: {
        functions: (oI) => ({
            ...oI,
            getMFARequirementsForAuth: async () => [
                {
                    oneOf: [
                        ${factorIds.join(",\n                        ")}
                    ],
                },
            ],
            // NOTE: Removing getRequiredSecondaryFactorsForUser override based on user feedback
            // getRequiredSecondaryFactorsForUser: async () => {
            //     return [${factorIds.join(", ")}];
            // },
        }),
    }`;
            }
        }

        init += `
})`;
        return init;
    },
    accountLinking: (hasMFA: boolean) => {
        // Reverting to returning plain object to avoid constructor/reference errors at runtime,
        // even though SDK might expect class instance. Underlying "not allowed to" error likely needs SDK fix.
        const shouldRequireVerification = hasMFA;
        return `AccountLinking.init({
            shouldDoAutomaticAccountLinking: async (
                newAccountInfo: AccountInfoWithRecipeId, // Corrected type
                user: User | undefined, // Use undefined directly instead of Optional<User>
                session: any, // Use any for session as SessionContainer import was removed
                tenantId: string,
                userContext: any // Use any for userContext instead of Dict<string, Any>
            ) => {
                // TODO: Add custom logic here based on your requirements
                return { // Return plain object
                    shouldAutomaticallyLink: true,
                    shouldRequireVerification: ${shouldRequireVerification}
                };
            }
        })`;
    },
    emailVerification: () => `EmailVerification.init({
        mode: "REQUIRED" // Revert back to REQUIRED
    })`,
    totp: () => `TOTP.init()`,
    multitenancy: () => `Multitenancy.init({
        override: {
            functions: (oI) => {
                return {
                    ...oI,
                    // Add any necessary overrides for Multitenancy + MFA interaction here
                };
            },
        },
    })`,
} as const;

export const generateTypeScriptTemplate = (
    {
        configType,
        userArguments,
        isFullStack = false, // Default to false
    }: TypeScriptTemplateOptions,
    _framework?: string // Keep framework for potential future use or logging
): string => {
    const recipes = configToRecipes[configType];
    const hasMFA =
        configType === "multifactorauth" || (userArguments?.secondfactors && userArguments.secondfactors.length > 0);

    // Add MFA recipes if needed
    if (hasMFA && !recipes.includes("multiFactorAuth")) {
        recipes.push("multiFactorAuth");
    }
    if (hasMFA && userArguments?.secondfactors?.includes("totp") && !recipes.includes("totp")) {
        recipes.push("totp");
    }
    if (hasMFA && !recipes.includes("emailVerification")) {
        recipes.push("emailVerification");
    }
    if (hasMFA && !recipes.includes("accountLinking")) {
        recipes.push("accountLinking");
    }

    // Add Passwordless recipe if any OTP/Link factors are used (first or second)
    const hasPasswordlessFactor =
        userArguments?.firstfactors?.some((f) => f.startsWith("otp-") || f.startsWith("link-")) ||
        userArguments?.secondfactors?.some((f) => f.startsWith("otp-") || f.startsWith("link-"));

    if (hasPasswordlessFactor && !recipes.includes("passwordless")) {
        recipes.push("passwordless");
    }

    // Get the correct appInfo based on fullstack context
    const appInfo = getAppInfo(isFullStack);

    // Generate imports
    let imports = recipes
        .map((recipe) => tsRecipeImports[recipe as keyof typeof tsRecipeImports])
        .filter(Boolean)
        .join("\n");

    // Add imports for types used in overrides
    if (recipes.includes("accountLinking")) {
        // Correct type name and remove problematic SessionContainer import path
        imports += `\nimport type { AccountInfoWithRecipeId } from "supertokens-node/recipe/accountlinking/types";`; // Corrected type name
        imports += `\nimport type { User } from "supertokens-node/types";`;
    }

    // Removed incorrect import from "typing"

    // Generate recipe initializations using a switch for clarity and type safety
    // Ensure Session is handled last
    const initRecipes = recipes.filter((r) => r !== "session");
    const sessionInitFunc = tsRecipeInits.session; // Assuming session init takes no args

    const recipeInits = initRecipes
        .map((recipe) => {
            switch (recipe) {
                case "thirdParty":
                    return tsRecipeInits.thirdParty(oAuthProviders);
                case "multiFactorAuth":
                    return tsRecipeInits.multiFactorAuth(userArguments?.firstfactors, userArguments?.secondfactors);
                case "accountLinking":
                    return tsRecipeInits.accountLinking(hasMFA ?? false); // Pass hasMFA, default to false if undefined
                case "passwordless":
                case "passwordless":
                    return tsRecipeInits.passwordless(userArguments);
                case "emailVerification": // Reverted: EV init takes no args now
                    return tsRecipeInits.emailVerification();
                // Add cases for other recipes that might need specific arguments in the future
                case "emailPassword":
                case "dashboard":
                case "userRoles":
                // case "emailVerification": // Handled separately below
                case "totp":
                case "multitenancy":
                    // Call recipes that don't require arguments (excluding EV)
                    const initFunc = tsRecipeInits[recipe];
                    if (typeof initFunc === "function") {
                        // We check if it's a function before calling
                        // Need to cast to any because TS struggles with indexed access signature inference here
                        return (initFunc as any)();
                    }
                    console.warn(`No initializer function found for recipe: ${recipe}`);
                    return null;
                default:
                    console.warn(`Unknown recipe encountered: ${recipe}`);
                    return null;
            }
        })
        .filter(Boolean); // Get array of init strings

    // Add Session.init() at the end
    recipeInits.push(sessionInitFunc());

    // Removed unused recipeInitsString variable

    // Construct the final template string
    let template =
        imports +
        "\n" +
        // Reverting to value import for TypeInput based on user's working config
        `import type { TypeInput } from "supertokens-node/types";

export function getApiDomain() {
    const apiPort = ${appInfo.defaultApiPort};
    const apiUrl = \`http://localhost:\${apiPort}\`;
    return apiUrl;
}

export function getWebsiteDomain() {
    const websitePort = ${appInfo.defaultWebsitePort};
    const websiteUrl = \`http://localhost:\${websitePort}\`;
    return websiteUrl;
}` +
        "\n";

    // Add configuration
    template += `\nexport const SuperTokensConfig: TypeInput = {
    supertokens: {
        connectionURI: "${config.connectionURI}",
    },
    appInfo: {
        appName: "${appInfo.appName}",
        apiDomain: getApiDomain(),
        websiteDomain: getWebsiteDomain(),
        apiBasePath: "${appInfo.apiBasePath}", // Use correct base path
        websiteBasePath: "${appInfo.websiteBasePath}", // Use correct base path
    },
    recipeList: [
        ${recipeInits.join(",\n        ")}
    ],
};`;

    return template;
};
