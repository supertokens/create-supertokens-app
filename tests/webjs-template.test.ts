import { describe, it, expect } from "vitest";
import { generateWebJSTemplate } from "../boilerplate/frontend/shared/web-js/template"; // Adjust path
import { UserFlags } from "../lib/ts/types"; // Adjust path

describe("WebJS Template Generation", () => {
    it("should generate correct config for EmailPassword only (Angular)", () => {
        const args = {
            configType: "emailpassword" as const,
            userArguments: { manager: "npm" } as UserFlags,
            framework: "angular",
        };
        const result = generateWebJSTemplate(args);
        expect(result).toMatchSnapshot();
    });

    it("should generate correct config for ThirdParty only (Vue)", () => {
        const args = {
            configType: "thirdparty" as const,
            userArguments: { manager: "npm" } as UserFlags,
            framework: "vue",
        };
        const result = generateWebJSTemplate(args);
        expect(result).toMatchSnapshot();
    });

    it("should generate correct config for EmailPassword + ThirdParty + OTP Email MFA (Solid)", () => {
        const args = {
            configType: "thirdpartyemailpassword" as const,
            userArguments: {
                manager: "npm" as const,
                firstfactors: ["emailpassword", "thirdparty"],
                secondfactors: ["otp-email"],
            } as UserFlags,
            framework: "solid",
        };
        const result = generateWebJSTemplate(args);
        // Check for key MFA/EV inclusions
        expect(result).toContain("supertokensUIMultiFactorAuth.init");
        expect(result).toContain("supertokensUIEmailVerification.init");
        expect(result).toContain("MultiFactorAuth.init()"); // Core SDK init
        expect(result).toContain("EmailVerification.init()"); // Core SDK init
        expect(result).toMatchSnapshot();
    });

    it("should generate correct config for EmailPassword + OTP Phone/Email (Angular)", () => {
        const args = {
            configType: "emailpassword" as const, // configType might be inferred differently
            userArguments: {
                manager: "npm" as const,
                firstfactors: ["emailpassword", "otp-phone", "otp-email"],
            } as UserFlags,
            framework: "angular",
        };
        const result = generateWebJSTemplate(args);
        // Check specific parts for passwordless config
        expect(result).toContain('contactMethod: "EMAIL_OR_PHONE"');
        expect(result).toContain('flowType: "USER_INPUT_CODE"'); // Only OTP factors
        expect(result).toMatchSnapshot();
    });

    // TODO: Add more specific test cases for different factor combinations and frameworks
});
