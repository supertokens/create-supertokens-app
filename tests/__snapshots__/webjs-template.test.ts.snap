// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WebJS Template Generation > should generate correct config for EmailPassword + OTP Phone/Email (Angular) 1`] = `
"import SuperTokens from "supertokens-web-js";
import Session from "supertokens-web-js/recipe/session";

export function getApiDomain() {
    const apiPort = 3001;
    const apiUrl = \`http://localhost:\${apiPort}\`;
    return apiUrl;
}

export function getWebsiteDomain() {
    const websitePort = 3000;
    const websiteUrl = \`http://localhost:\${websitePort}\`;
    return websiteUrl;
}

export function initSuperTokensUI() {
    (window as any).supertokensUIInit("supertokensui", {
        appInfo: {
            // Call generated helper functions
            websiteDomain: getWebsiteDomain(),
            apiDomain: getApiDomain(),
            appName: "SuperTokens Demo App", // Keep using direct value for appName
        },
        recipeList: [
            (window as any).supertokensUISession.init(),
            (window as any).supertokensUIEmailPassword.init(),
            (window as any).supertokensUIPasswordless.init({
                contactMethod: "EMAIL_OR_PHONE",
                flowType: "USER_INPUT_CODE"
            })
        ],
    });
}

export function initSuperTokensWebJS() {
    SuperTokens.init({
        appInfo: {
            appName: "SuperTokens Demo App", // Use dynamic appName
            apiDomain: getApiDomain(), // Call helper function
        },
        recipeList: [Session.init()]
    });
}"
`;

exports[`WebJS Template Generation > should generate correct config for EmailPassword + ThirdParty + OTP Email MFA (Solid) 1`] = `
"import SuperTokens from "supertokens-web-js";
import Session from "supertokens-web-js/recipe/session";
import EmailVerification from "supertokens-web-js/recipe/emailverification";
import MultiFactorAuth from "supertokens-web-js/recipe/multifactorauth";

export function getApiDomain() {
    const apiPort = 3001;
    const apiUrl = \`http://localhost:\${apiPort}\`;
    return apiUrl;
}

export function getWebsiteDomain() {
    const websitePort = 3000;
    const websiteUrl = \`http://localhost:\${websitePort}\`;
    return websiteUrl;
}

export function initSuperTokensUI() {
    (window as any).supertokensUIInit("supertokensui", {
        appInfo: {
            // Call generated helper functions
            websiteDomain: getWebsiteDomain(),
            apiDomain: getApiDomain(),
            appName: "SuperTokens Demo App", // Keep using direct value for appName
        },
        recipeList: [
            (window as any).supertokensUISession.init(),
            (window as any).supertokensUIEmailPassword.init(),
            (window as any).supertokensUIThirdParty.init({
                signInAndUpFeature: {
                    providers: [
                        (window as any).supertokensUIThirdParty.Github.init(),
                        (window as any).supertokensUIThirdParty.Google.init(),
                        (window as any).supertokensUIThirdParty.Apple.init(),
                        (window as any).supertokensUIThirdParty.Twitter.init(),
                    ],
                },
            }),
            (window as any).supertokensUIPasswordless.init({
                contactMethod: "EMAIL",
                flowType: "USER_INPUT_CODE"
            }),
            (window as any).supertokensUIMultiFactorAuth.init({
        firstFactors: ["emailpassword", "thirdparty"],
        override: {
            functions: (originalImplementation) => ({
                ...originalImplementation,
                getMFARequirementsForAuth: () => [
                    {
                        oneOf: [
                            (window as any).supertokensUIMultiFactorAuth.FactorIds.OTP_EMAIL
                        ],
                    },
                ],
                // Override the getRequiredSecondaryFactorsForUser function to ensure
                // the user is required to set up the second factors
                getRequiredSecondaryFactorsForUser: async () => {
                    return [(window as any).supertokensUIMultiFactorAuth.FactorIds.OTP_EMAIL];
                },
            }),
        }
    }),
            (window as any).supertokensUIEmailVerification.init({
        mode: "REQUIRED",
    })
        ],
    });
}

export function initSuperTokensWebJS() {
    SuperTokens.init({
        appInfo: {
            appName: "SuperTokens Demo App", // Use dynamic appName
            apiDomain: getApiDomain(), // Call helper function
        },
        // Core SDK init only needs Session, EmailVerification, and MultiFactorAuth for MFA
        recipeList: [Session.init(), EmailVerification.init(), MultiFactorAuth.init()]
    });
}"
`;

exports[`WebJS Template Generation > should generate correct config for EmailPassword only (Angular) 1`] = `
"import SuperTokens from "supertokens-web-js";
import Session from "supertokens-web-js/recipe/session";

export function getApiDomain() {
    const apiPort = 3001;
    const apiUrl = \`http://localhost:\${apiPort}\`;
    return apiUrl;
}

export function getWebsiteDomain() {
    const websitePort = 3000;
    const websiteUrl = \`http://localhost:\${websitePort}\`;
    return websiteUrl;
}

export function initSuperTokensUI() {
    (window as any).supertokensUIInit("supertokensui", {
        appInfo: {
            // Call generated helper functions
            websiteDomain: getWebsiteDomain(),
            apiDomain: getApiDomain(),
            appName: "SuperTokens Demo App", // Keep using direct value for appName
        },
        recipeList: [
            (window as any).supertokensUISession.init(),
            (window as any).supertokensUIEmailPassword.init()
        ],
    });
}

export function initSuperTokensWebJS() {
    SuperTokens.init({
        appInfo: {
            appName: "SuperTokens Demo App", // Use dynamic appName
            apiDomain: getApiDomain(), // Call helper function
        },
        recipeList: [Session.init()]
    });
}"
`;

exports[`WebJS Template Generation > should generate correct config for ThirdParty only (Vue) 1`] = `
"import SuperTokens from "supertokens-web-js";
import Session from "supertokens-web-js/recipe/session";

export function getApiDomain() {
    const apiPort = 3001;
    const apiUrl = \`http://localhost:\${apiPort}\`;
    return apiUrl;
}

export function getWebsiteDomain() {
    const websitePort = 3000;
    const websiteUrl = \`http://localhost:\${websitePort}\`;
    return websiteUrl;
}

export function initSuperTokensUI() {
    (window as any).supertokensUIInit("supertokensui", {
        appInfo: {
            // Call generated helper functions
            websiteDomain: getWebsiteDomain(),
            apiDomain: getApiDomain(),
            appName: "SuperTokens Demo App", // Keep using direct value for appName
        },
        recipeList: [
            (window as any).supertokensUISession.init(),
            (window as any).supertokensUIThirdParty.init({
                signInAndUpFeature: {
                    providers: [
                        (window as any).supertokensUIThirdParty.Github.init(),
                        (window as any).supertokensUIThirdParty.Google.init(),
                        (window as any).supertokensUIThirdParty.Apple.init(),
                        (window as any).supertokensUIThirdParty.Twitter.init(),
                    ],
                },
            })
        ],
    });
}

export function initSuperTokensWebJS() {
    SuperTokens.init({
        appInfo: {
            appName: "SuperTokens Demo App", // Use dynamic appName
            apiDomain: getApiDomain(), // Call helper function
        },
        recipeList: [Session.init()]
    });
}"
`;
